### PR TITLE
New system table zookeeper connection

### PR DESCRIPTION
**Pull request by mateng0915**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added `system.zookeeper_connection` table that shows information about ZooKeeper connections

Description
=========
Currently if we want to check our CK instance connected to which zk,
we need login to CK deployed machine then using like lsof -iTCP -NP these kind of linux command
and it is not easy to use, if we want to check all Cluster instance connected information

Solution
======
Added a new system table, which table save 
Connected zk ip:port
connected zk index number (which was in metrika.xml or config.xml)

With using the clusterAllReplicas() functions, only need executed one simple sql cmd
will know whole CK cluster's all CK server connected info with zookeeper

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by tavplubix** [comment]
I would prefer to implement it as a function, not as a system table because such a table would always contain only one row.

**Comment by tavplubix** [comment]
Well, multiple connections may exist simultaneously with the "auxiliary zookeeper" feature, but a function can take one optional argument: the name of an auxiliary zookeeper.

**Comment by UnamedRus** [comment]
May be, Zookeeper requests in fly?

**Comment by FrankChen021** [comment]
This is a very help feature. I'm wondering if we can add one more field like `connected_time` to show the time when the connection is established. This helps us know how when a connection is aborted and re-established.

**Comment by mateng0915** [comment]
> This is a very help feature. I'm wondering if we can add one more field like `connected_time` to show the time when the connection is established. This helps us know how when a connection is aborted and re-established.

That a very useful suggestion, will added one more column that shows the establish time

**Comment by mateng0915** [comment]
> 
Agree your opinion

**Comment by mateng0915** [comment]
> I would prefer to implement it as a function, not as a system table because such a table would always contain only one row.

I think the system table is better than function, because for checking the zk connection address, check the whole cluster's all servers' connection is also very important, if we have one system table, then can using FQDN() and clusterAllReplicas() functions to query all servers's zk address
eg: select FQDN() as host, zk_address  from clusterAllReplicas($cluster_name, system,zookeeper_connection) group by host, zk_address order by host;

**Comment by xiedeyantu** [comment]
I think system table is a better way, this can record some zk connection errors.

**Comment by mateng0915** [comment]
> I would prefer to implement it as a function, not as a system table because such a table would always contain only one row.

actually it is not only one row, because clickhouse support multi-zookeeper cluster features, (**auxiliary zookeeper**), and in our CK Cluster the biggest storage cluster has three zk clusters 

Eg, here is the example when one CK cluster has two zk cluster
image

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 24ba864a91a997b78e39136286614b5eaafae7a4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/45245/24ba864a91a997b78e39136286614b5eaafae7a4/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

**Comment by mateng0915** [comment]
> > I would prefer to implement it as a function, not as a system table because such a table would always contain only one row.
> 
> actually it is not only one row, because clickhouse support multi-zookeeper cluster features, (**auxiliary zookeeper**), and in our CK Cluster the biggest storage cluster has three zk clusters
> 
> Eg, here is the example when one CK cluster has two zk cluster image

pin @tavplubix

File: src/Interpreters/Context.h [hunk 1]
```diff
@@ -720,6 +720,8 @@ class Context: public std::enable_shared_from_this<Context>
     const IUserDefinedSQLObjectsLoader & getUserDefinedSQLObjectsLoader() const;
     IUserDefinedSQLObjectsLoader & getUserDefinedSQLObjectsLoader();
     void loadOrReloadUserDefinedExecutableFunctions(const Poco::Util::AbstractConfiguration & config);
+    void loadOrReloadModels(const Poco::Util::AbstractConfiguration & config);
```
Comment (tavplubix) [inline src/Interpreters/Context.h:723]: What is `loadOrReloadModels`?

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+struct ContextSharedPart;
+
+NamesAndTypesList StorageSystemZooKeeperConnection::getNamesAndTypes()
+{
+    return {
+        {"zookeeper_name", std::make_shared<DataTypeString>()},
+        {"zookeeper_address", std::make_shared<DataTypeString>()},
+        {"zookeeper_index", std::make_shared<DataTypeUInt8>()},
+        {"zookeeper_connected_time", std::make_shared<DataTypeDateTime>()},
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:18]: Consider removing `zookeeper_` prefix

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace DB
+{
+struct ContextSharedPart;
+
+NamesAndTypesList StorageSystemZooKeeperConnection::getNamesAndTypes()
+{
+    return {
+        {"zookeeper_name", std::make_shared<DataTypeString>()},
+        {"zookeeper_address", std::make_shared<DataTypeString>()},
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:16]: It makes sense to add `port` column as well

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        {"zookeeper_index", std::make_shared<DataTypeUInt8>()},
+        {"zookeeper_connected_time", std::make_shared<DataTypeDateTime>()},
+    };
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:19]: Columns showing `ZooKeeper::expired()`, `ZooKeeper::getApiVersion()`, and `ZooKeeper::getClientID()` will be useful as well

**Review by tavplubix — COMMENTED** [review]
Also please add a functional test

**Comment by mateng0915** [comment]
> Also please add a functional test
Many thx your review comments, very helpful
and now I finish added test case and fix review comments. Please kindly help review the PR, thx

File: tests/queries/0_stateless/02731_system_zookeeper_connection.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+DROP TABLE IF EXISTS test_zk_connection_table;
+
+CREATE TABLE test_zk_connection_table (
+    key UInt64
+)
+ENGINE ReplicatedMergeTree('/clickhouse/{database}/02731_zk_connection/{shard}', '{replica}')
+ORDER BY tuple();
+
+select count(*)  from system.zookeeper_connection where name='default_zookeeper';
```
Comment (tavplubix) [inline tests/queries/0_stateless/02731_system_zookeeper_connection.sql:11]: We can select some columns like host, port, is_expired (it's expected to print something like `localhost, 9181, 0`)

File: src/Storages/System/StorageSystemZooKeeperConnection.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    res_columns[3]->insert(context->getZooKeeper()->getConnectedZooKeeperIndex());
+    res_columns[4]->insert(context->getZooKeeperSessionUptime());
+    res_columns[5]->insert(context->getZooKeeper()->expired());
+    res_columns[6]->insert(context->getZooKeeper()->getApiVersion());
+    res_columns[7]->insert(context->getZooKeeper()->getClientID());
+
+    std::map<String, zkutil::ZooKeeperPtr>::iterator iter;
+
+    for (iter = context->getAuxiliaryZooKeepers().begin(); iter !=
+         context->getAuxiliaryZooKeepers().end(); iter ++)
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeperConnection.cpp:40]:
```suggestion
    std::map<String, zkutil::ZooKeeperPtr>::iterator iter;

    for (const auto & elem : context->getAuxiliaryZooKeepers())
```

**Review by tavplubix — APPROVED** [review]

File: tests/queries/0_stateless/02731_system_zookeeper_connection.sql (hunk 5 above)
Comment (mateng0915) [reply tests/queries/0_stateless/02731_system_zookeeper_connection.sql:11]: correct, can select theses

File: tests/queries/0_stateless/02731_system_zookeeper_connection.sql [hunk 7]
```diff
@@ -0,0 +1,13 @@
+-- Tags: system_zookeeper_connection table
+
```
[the remark this answers is not in the document]
Comment (mateng0915) [reply tests/queries/0_stateless/02731_system_zookeeper_connection.sql]: > This tag is not needed, see https://clickhouse.com/docs/en/development/tests#restricting-test-runs

Thx your  comments, I finished the review comments, please kindly help to review it . Thx

**Comment by tavplubix** [comment]
Please check failed tests:
Fast test — fail: 1, passed: 4331, skipped: 921 
Style Check — Check style failed.

**Comment by mateng0915** [comment]
> Please check failed tests: Fast test — fail: 1, passed: 4331, skipped: 921 Style Check — Check style failed.

Thx remind, fixed the failure

**Comment by tavplubix** [comment]
Merging master and restarting CI will not help, failed tests have to be fixed. The new test fails like this:
```
2023-05-10 06:42:00 02731_system_zookeeper_connection:                                      [ FAIL ] 0.24 sec. - result differs with reference: 
2023-05-10 06:42:00 --- /usr/share/clickhouse-test/queries/0_stateless/02731_system_zookeeper_connection.reference	2023-05-10 06:39:35.269198586 -0400
2023-05-10 06:42:00 +++ /tmp/clickhouse-test/0_stateless/02731_system_zookeeper_connection.stdout	2023-05-10 06:42:00.497852893 -0400
2023-05-10 06:42:00 @@ -1 +1 @@
2023-05-10 06:42:00 -127.0.0.1	9181	0
2023-05-10 06:42:00 +[	:1]:9181	0
2023-05-10 06:42:00 
2023-05-10 06:42:00 
```

It's because the server configurations for Fast test and other Stateless tests checks are a bit different. Possible solutions:
 - try using the same configuration
 - update the reference file and add `-- Tags: no-fasttest` comment to the first line of the test

**Comment by tavplubix** [comment]
Merging master and restarting CI will not help, failed tests have to be fixed. The new test fails like this:
```
2023-05-10 06:42:00 02731_system_zookeeper_connection:                                      [ FAIL ] 0.24 sec. - result differs with reference: 
2023-05-10 06:42:00 --- /usr/share/clickhouse-test/queries/0_stateless/02731_system_zookeeper_connection.reference	2023-05-10 06:39:35.269198586 -0400
2023-05-10 06:42:00 +++ /tmp/clickhouse-test/0_stateless/02731_system_zookeeper_connection.stdout	2023-05-10 06:42:00.497852893 -0400
2023-05-10 06:42:00 @@ -1 +1 @@
2023-05-10 06:42:00 -127.0.0.1	9181	0
2023-05-10 06:42:00 +[	:1]:9181	0
2023-05-10 06:42:00 
2023-05-10 06:42:00 
```

It's because the server configurations for Fast test and other Stateless tests checks are a bit different. Possible solutions:
 - try using the same configuration
 - update the reference file and add `-- Tags: no-fasttest` comment to the first line of the test

**Comment by mateng0915** [comment]
> Merging master and restarting CI will not help, failed tests have to be fixed. The new test fails like this:
> 
> ```
> 2023-05-10 06:42:00 02731_system_zookeeper_connection:                                      [ FAIL ] 0.24 sec. - result differs with reference: 
> 2023-05-10 06:42:00 --- /usr/share/clickhouse-test/queries/0_stateless/02731_system_zookeeper_connection.reference	2023-05-10 06:39:35.269198586 -0400
> 2023-05-10 06:42:00 +++ /tmp/clickhouse-test/0_stateless/02731_system_zookeeper_connection.stdout	2023-05-10 06:42:00.497852893 -0400
> 2023-05-10 06:42:00 @@ -1 +1 @@
> 2023-05-10 06:42:00 -127.0.0.1	9181	0
> 2023-05-10 06:42:00 +[	:1]:9181	0
> 2023-05-10 06:42:00 
> 2023-05-10 06:42:00 
> ```
> 
> It's because the server configurations for Fast test and other Stateless tests checks are a bit different. Possible solutions:
> 
> * try using the same configuration
> * update the reference file and add `-- Tags: no-fasttest` comment to the first line of the test

Thx , Will added Tags to solve it

**Comment by mateng0915** [comment]
> Merging master and restarting CI will not help, failed tests have to be fixed. The new test fails like this:
> 
> ```
> 2023-05-10 06:42:00 02731_system_zookeeper_connection:                                      [ FAIL ] 0.24 sec. - result differs with reference: 
> 2023-05-10 06:42:00 --- /usr/share/clickhouse-test/queries/0_stateless/02731_system_zookeeper_connection.reference	2023-05-10 06:39:35.269198586 -0400
> 2023-05-10 06:42:00 +++ /tmp/clickhouse-test/0_stateless/02731_system_zookeeper_connection.stdout	2023-05-10 06:42:00.497852893 -0400
> 2023-05-10 06:42:00 @@ -1 +1 @@
> 2023-05-10 06:42:00 -127.0.0.1	9181	0
> 2023-05-10 06:42:00 +[	:1]:9181	0
> 2023-05-10 06:42:00 
> 2023-05-10 06:42:00 
> ```
> 
> It's because the server configurations for Fast test and other Stateless tests checks are a bit different. Possible solutions:
> 
> * try using the same configuration
> * update the reference file and add `-- Tags: no-fasttest` comment to the first line of the test

Stateless test Failure Fixed, Thx your info about how to handle Fast test and Stateless test

**Comment by mateng0915** [comment]
> > Merging master and restarting CI will not help, failed tests have to be fixed. The new test fails like this:
> > ```
> > 2023-05-10 06:42:00 02731_system_zookeeper_connection:                                      [ FAIL ] 0.24 sec. - result differs with reference: 
> > 2023-05-10 06:42:00 --- /usr/share/clickhouse-test/queries/0_stateless/02731_system_zookeeper_connection.reference	2023-05-10 06:39:35.269198586 -0400
> > 2023-05-10 06:42:00 +++ /tmp/clickhouse-test/0_stateless/02731_system_zookeeper_connection.stdout	2023-05-10 06:42:00.497852893 -0400
> > 2023-05-10 06:42:00 @@ -1 +1 @@
> > 2023-05-10 06:42:00 -127.0.0.1	9181	0
> > 2023-05-10 06:42:00 +[	:1]:9181	0
> > 2023-05-10 06:42:00 
> > 2023-05-10 06:42:00 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It's because the server configurations for Fast test and other Stateless tests checks are a bit different. Possible solutions:
> > 
> > * try using the same configuration
> > * update the reference file and add `-- Tags: no-fasttest` comment to the first line of the test
> 
> Thx , Will added Tags to solve it

stateless test case failure fixed

**Comment by tavplubix** [comment]
Integration tests (release) [2/4] - `test_storage_mysql/test.py::test_table_function` is a bit flaky, unrelated
Stress test (debug) - https://github.com/ClickHouse/ClickHouse/pull/49673, unrelated

Docs Check Pending — expect adding docs for pr-feature - @mateng0915, could you please add documentation for the new table?

**Comment by mateng0915** [comment]
> Integration tests (release) [2/4] - `test_storage_mysql/test.py::test_table_function` is a bit flaky, unrelated Stress test (debug) - #49673, unrelated
> 
> Docs Check Pending — expect adding docs for pr-feature - @mateng0915, could you please add documentation for the new table?
ok I will add a doc about new system table

**Comment by mateng0915** [comment]
> Integration tests (release) [2/4] - `test_storage_mysql/test.py::test_table_function` is a bit flaky, unrelated Stress test (debug) - #49673, unrelated
> 
> Docs Check Pending — expect adding docs for pr-feature - @mateng0915, could you please add documentation for the new table?

Added Doc and pass the Docs Check
Thx review

File: docs/en/operations/system-tables/zookeeper_connection.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+---
+slug: /zh/operations/system-tables/zookeeper_connection
+---
+#zookeeper_connection
+
+This table does not exist if ClickHouse Cluster does not config zookeeper. The 'system.zookeeper_connection' table will shows current clickhouse-cluster connected zookeepers info.
+
+each rows will shows the node information of a zk cluster connected to.
```
Comment (tavplubix) [inline docs/en/operations/system-tables/zookeeper_connection.md:8]:
```suggestion
This table does not exist if ZooKeeper is not configured. The 'system.zookeeper_connection' table shows current connections to ZooKeeper (including auxiliary ZooKeepers). Each row shows information about one connection.
```

File: docs/en/operations/system-tables/zookeeper_connection.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+Columns:
+
+-   `name` ([String](../../sql-reference/data-types/string.md)) — zookeeper cluster name.
+-   `host` ([String](../../sql-reference/data-types/string.md)) — connected zookeeper node host.
+-   `port` ([String](../../sql-reference/data-types/string.md)) — connected zookeeper node port.
+-   `index` ([UInt8](../../sql-reference/data-types/int-uint.md)) — connected zookeeper node index in zk cluster config.
+-   `connected_time` ([String](../../sql-reference/data-types/string.md)) — when establish the connection to this zk node.
+-   `is_expired` ([UInt8](../../sql-reference/data-types/int-uint.md)) — Does current connection expired.
+-   `keeper_api_version` ([String](../../sql-reference/data-types/string.md)) — shows keeper api version.
+-   `client_id` ([UInt64](../../sql-reference/data-types/int-uint.md)) — zookeeper connection session id.
```
Comment (tavplubix) [inline docs/en/operations/system-tables/zookeeper_connection.md:19]:
```suggestion
-   `name` ([String](../../sql-reference/data-types/string.md)) — ZooKeeper cluster's name.
-   `host` ([String](../../sql-reference/data-types/string.md)) — The hostname/IP of the ZooKeeper node that ClickHouse connected to.
-   `port` ([String](../../sql-reference/data-types/string.md)) — The port of the ZooKeeper node that ClickHouse connected to.
-   `index` ([UInt8](../../sql-reference/data-types/int-uint.md)) — The index of the ZooKeeper node that ClickHouse connected to. The index is from ZooKeeper config.
-   `connected_time` ([String](../../sql-reference/data-types/string.md)) — When the connection was established
-   `is_expired` ([UInt8](../../sql-reference/data-types/int-uint.md)) — Is the current connection expired.
-   `keeper_api_version` ([String](../../sql-reference/data-types/string.md)) — Keeper API version.
-   `client_id` ([UInt64](../../sql-reference/data-types/int-uint.md)) — Session id of the connection.
```

**Review by tavplubix — APPROVED** [review]

File: docs/en/operations/system-tables/zookeeper_connection.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
+---
+slug: /zh/operations/system-tables/zookeeper_connection
```
Comment (tavplubix) [inline docs/en/operations/system-tables/zookeeper_connection.md:2]:
```suggestion
slug: /en/operations/system-tables/zookeeper_connection
```

**Pull request merged**